packages/forms/backend/sirius-components-collaborative-forms/src/main/resources/schema/form.graphqls

```diff
@@ -498,3 +500,29 @@ type PushButtonSuccessPayload {
 }
 
 union PushButtonPayload = ErrorPayload | PushButtonSuccessPayload
+
+extend type EditingContext {
+  completionProposals(input: CompletionRequestInput!): CompletionRequestPayload!
+}
+
+input CompletionRequestInput {
+  id: ID!
+  editingContextId: ID!
+  representationId: ID!
+  widgetId: ID!
+  currentText: String!
+  cursorPosition: Int!
+}
+
+union CompletionRequestPayload = ErrorPayload | CompletionRequestSuccessPayload
+
+type CompletionRequestSuccessPayload {
+  id: ID!
+  proposals: [CompletionProposal!]!
+}
+
+type CompletionProposal {
+  description: String!
+  textToInsert: String!
+  charsToReplace: Int!
+}
```

Comment:
I wouldn't use this approach to plug the completion proposals in the GraphQL schema for a couple of reasons:

- The input / payload pattern is used for mutations, you don't need it here since you are not mutating anything on the backend. You do not need an error payload here either. We don't have any for our queries and we have a lot of queries.
- You may need an internal Input / Payload to compute the data in the editing context but that's an internal concern that does not need to appear for consumers of the API (for example by requiring them to provide an UUID id)
- You are adding a very specific field on the editing context while we have made sure instead to keep dedicated fiels like that on the representation descriptions instead. Have a look at the way we compute, the palette, the initial label or the context of the connector tool.
- Since you have this new field all the one on the top of the editing context and since it's an input, you have to give it some parameters that are not required such as id, editingContextId and the representation id. Speaking about the editingContextId, with your approach you need to give it twice (once to find the editing context object and another time as the input). We solved the issue for both the editing context id and the representation id already to compute those kinds of fields for diagrams (for example for the palette). For example with the following query:

```
  query getToolSections($editingContextId: ID!, $diagramId: ID!, $diagramElementId: ID!) {
    viewer {
      editingContext(editingContextId: $editingContextId) {
        representation(representationId: $diagramId) {
          description {
            ... on DiagramDescription {
              toolSections(diagramElementId: $diagramElementId) {
                label
                // Additional fields to compute the palette
              }
            }
          }
        }
      }
    }
  }
```

We have thus setup everything to propagate both the editing context id and the representation id to all child data fetchers recursively, see the usage of `LocalContextConstants` on the backend. Your query should instead be:


```
  query completionProposals($editingContextId: ID!, $formId: ID!, $widgetId: ID!, $currentText: String!, $cursorPosition: Int!) {
    viewer {
      editingContext(editingContextId: $editingContextId) {
        representation(representationId: $formId) {
          description {
            ... on FormDescription {
              completionProposals(widgetId: $widgetId, currentText: $currentText, cursorPosition: $cursorPosition) {
                // Specific fields
              }
            }
          }
        }
      }
    }
  }
```

We have successfully applied this pattern on ContextMenu.tsx, ContextualPalette.tsx and EditLabelUIWithInitialContent.ts for example.

Reply:
Hmm, injecting `IEditingContextEventProcessorRegistry` into `RepresentationService` create a dependency cycle.

```
The dependencies of some of the beans in the application context form a cycle:

   editingContextActionsDataFetcher defined in file [sirius-components/packages/sirius-web/backend/sirius-web-graphql/target/classes/org/eclipse/sirius/web/graphql/datafetchers/editingcontext/EditingContextActionsDataFetcher.class]
┌─────┐
|  editingContextEventProcessorRegistry defined in file [sirius-components/packages/core/backend/sirius-components-collaborative/target/classes/org/eclipse/sirius/components/collaborative/editingcontext/EditingContextEventProcessorRegistry.class]
↑     ↓
|  editingContextEventProcessorFactory defined in file [sirius-components/packages/core/backend/sirius-components-collaborative/target/classes/org/eclipse/sirius/components/collaborative/editingcontext/EditingContextEventProcessorFactory.class]
↑     ↓
|  deleteRepresentationEventHandler defined in file [sirius-components/packages/sirius-web/backend/sirius-web-services/target/classes/org/eclipse/sirius/web/services/representations/DeleteRepresentationEventHandler.class]
↑     ↓
|  representationService defined in file [sirius-components/packages/sirius-web/backend/sirius-web-services/target/classes/org/eclipse/sirius/web/services/representations/RepresentationService.class]
└─────┘
```

Reply:
Not being able to handle in memory representations would indeed be an issue. We could modify the support of the deletion of representations by simply making DeleteRepresentationEventHandler depend on IRepresentationRepository directly because RepresentationService#delete is completely useless. It's only used at one location and it only hides one line of code.

Reply:
Now that I think about it since tou have the id of the representation can't we tell that this representation is one of the detail view?

Reply:
I don't think so, `PropertiesConfiguration` computes the id like this:
```javadoc
this.formId = UUID.nameUUIDFromBytes((PROPERTIES_PREFIX + objectIds).getBytes()).toString();
```